### PR TITLE
ROX-9944 / ROX-10648: Collect metrics from secured clusters

### PR DESCRIPTION
**Pull request by stehessel**
## Description

This PR adds a new `ClusterMetrics` message between Sensor and Central. It contains the number of nodes and the cpu capacity of a secured cluster. The idea is to i) display this information to users in the UI and ii) use these metrics as a basis for billing managed offerings.

The PR is split into two commits: The first one adds the metric collection on Sensor side and adds the gRPC pipeline. The second exposes the metrics to Prometheus via the `/metrics` endpoint in Central.

I also considered using Kubernetes events to trigger the messages based on changes, but ultimately I decided to go with a simple polling approach. There are two main reasons I favoured polling over the event based approach:
* Tracking changes via events can be very reliable, but it requires an immutable, replay-able event log to be robust. Otherwise, we run the risk of corrupting the metrics in case we miss an event. On the other hand, the polling approach is naturally self-healing.
* The polling load on the Kubernetes API is very small and the number of messages does not scale with the number of metrics polled. Conversely, when using event triggers, new metrics could significantly increase the number of messages.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] ~Evaluated and added CHANGELOG entry if required~
- [ ] ~Determined and documented upgrade steps~
- [ ] ~Documented user facing changes (create PR based on [openshift/openshift-docs]~(https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

* CI
* Tested that metrics show up in Prometheus when Central is deployed with monitoring support.

### CODE REVIEW

**Comment by stehessel** [comment]
@SimonBaeumer brought up an interesting point about how Sensor -> Central communication should be implemented: I addition to pipelines as used in this PR, there is also https://github.com/stackrox/stackrox/pull/219/files#diff-5eebef2d74ae81573c60ae033445bbc96d1d4fe325735839f47c945e6f5a1c97L210. @viswajithiii I'm wondering what the purpose of the pipelines is and when to use (or not to use) them.

**Comment by SimonBaeumer** [comment]
> @SimonBaeumer brought up an interesting point about how Sensor -> Central communication should be implemented: In addition to pipelines as used in this PR, there is also https://github.com/stackrox/stackrox/pull/219/files#diff-5eebef2d74ae81573c60ae033445bbc96d1d4fe325735839f47c945e6f5a1c97L210. @viswajithiii I'm wondering what the purpose of the pipelines is and when to use (or not to use) them.

@viswajithiii Stephan is referring to adding the message to the dispatching logic in [connection_impl.go](https://github.com/stackrox/stackrox/blob/747e03b381f99fe5bf437881af4579a50c82c943/central/sensor/service/connection/connection_impl.go#L212-L239).

File: proto/internalapi/central/cluster_metrics.proto [hunk 1]
```diff
@@ -0,0 +1,10 @@
+syntax = "proto3";
+
+option go_package = "central";
+
+package central;
+
+message ClusterMetrics {
+    int64 nodeCount = 1;
+    int64 cpuCapacity = 2;
```
Comment (viswajithiii) [inline proto/internalapi/central/cluster_metrics.proto:9]: Nit: this should be snake case according to our usual proto convention

File: sensor/kubernetes/clustermetrics/cluster_metrics.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+				} else {
+					log.Errorf("Collection of cluster metrics failed: %v", err.Error())
+				}
+			}
+		}
+	}()
+}
+
+func (cm *clusterMetricsImpl) collectMetrics() (*central.ClusterMetrics, error) {
+	nodes, err := cm.k8sClient.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
```
Comment (viswajithiii) [inline sensor/kubernetes/clustermetrics/cluster_metrics.go:93]: Use a context with a reasonable timeout

**Comment by viswajithiii** [comment]
> > @SimonBaeumer brought up an interesting point about how Sensor -> Central communication should be implemented: In addition to pipelines as used in this PR, there is also https://github.com/stackrox/stackrox/pull/219/files#diff-5eebef2d74ae81573c60ae033445bbc96d1d4fe325735839f47c945e6f5a1c97L210. @viswajithiii I'm wondering what the purpose of the pipelines is and when to use (or not to use) them.
> 
> @viswajithiii Stephan is referring to adding the message to the dispatching logic in [connection_impl.go](https://github.com/stackrox/stackrox/blob/747e03b381f99fe5bf437881af4579a50c82c943/central/sensor/service/connection/connection_impl.go#L212-L239).

I think the way you have it implemented is fine, and that ideally, more things would be implemented as pipelines rather than the ad hoc in-line logic in the switch.

**Comment by stehessel** [comment]
> One general concern I have is that this creates two different paths of metrics collection from sensors. Sensors already have their own metrics and prometheus which can be surfaced by scraping "/metrics" on the sensor pod, or by getting a diagnostic bundle. I wonder if we should unify these somehow?

> In any case, I understand these metrics are for billing, so I think these will have to be proactively piped to central regardless, but something to think about.

You are right, however as you point out these particular metrics need to be collected by Sensor and available in Central (or rather the data plane cluster in which Central is deployed). I also considered exposing the metrics in Sensor via its `/metrics` endpoint. We could then use an additional Prometheus instance - deployed in Sensor node - to scrape Sensor and federate the metrics to a Prometheus in the Central node. However, I think the solution in this PR is ultimately superior - for details see https://github.com/stackrox/architecture-decision-records/blob/main/ADR-0006-billing-metrics-flow.md.

**Pull request merged**